[Hern Ping] iP

hernpiblo opened the pull request:
# Pib

### *Pib* is a simple, lightweight and most importantly **FREE** software to track your tasks! 😎 

## What _Pib_ can do:

- ### Add tasks
  1. #### Todo : Normal tasks
      - `Todo read a book`
  2. #### Deadline : Tasks with a deadline
      - `Deadline SWE quiz /by 2021-09-01 2359`
  3. #### Events : Tasks with an event date
       - `Event hackathon /at 2021-10-10 1230`
- ### View entire list of tasks
- ### Mark tasks as completed
- ### Delete tasks
- ### Search for tasks
- ### Auto-save and load saved data

## Information

### View source code [here](https://github.com/hernpiblo/ip).
#### Snippet of code:
```java
public Pib() {
        this.ui = new Ui();
        this.list = new TaskList();
        try {
            this.storage = new Storage(DATA_FILE_PATH);
            storage.loadData(this.list);
        } catch (PibException e) {
            e.print();
        }
        this.parser = new Parser(this.list);
    }
```

## Get started
- [x] Download
- [x] Keep track of tasks
- [x] Enjoy
- [ ] Feedback

The source code is inspired by this quote:
> Keep it simple, stupid! - Kelly Johnson, lead engineer at the Lockheed Skunk Works

Commit:
Add Gradle support

hernpiblo committed:
Added Level 1 (Greet, Echo, Exit)

hernpiblo committed:
Added Level 2 (Add, List)

hernpiblo committed:
Added Level 3 (Mark as done)

hernpiblo committed:
Added Level 4 (ToDo, Event, Deadline)

hernpiblo committed:
Test fix for detached head

hernpiblo committed:
Merge commit '1db08c05315404b83b529059a5262c3970d42f24'

hernpiblo committed:
Added A-TextUiTesting, moved tasks into another package

hernpiblo committed:
Added A-TextUiTesting

hernpiblo committed:
Added Level 5 (Handle Errors)

hernpiblo committed:
Added Level 6 (Delete)

hernpiblo committed:
Added javadocs for public methods

hernpiblo committed:
Add date parser for Deadline and Event class

Date is treated as a normal string

Pib should be able to parse date and time so that it is able to
format it nicely for the list view

Let's encapsulate the function of parsing the date and time in the
Deadline and Event classes and add a public factory method to
process the raw string

hernpiblo committed:
Add read and write to file functionality

Pib resets the list everytime it is started

User has better experience if data can be saved and read
automatically

Let's add functionality to read from saved data and write any
changes to file

hernpiblo committed:
Merge branch 'branch-Level-8'

hernpiblo committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/Pib.java
#	src/main/java/Tasks/Todo.java

hernpiblo committed:
Refactor code into more classes for more OOP

Pib is having all the functionality in one class.

Having the functionalities split into classes follow OOP principles
better

Let's move Ui, Storage, TaskList and Parser into their own classes

hernpiblo committed:
Move classes into packages

Classes are all in the default package

Grouping classes in packages helps to standardize where similar
classes should be located

Let's create packages for some of these classes

hernpiblo committed:
Add JUnit tests

Methods are not being unit tested

Unit testing some public methods help to ensure that changes made
will not cause these methods to lose their functionality

Let's add JUnit tests for Deadline and Todo classes

hernpiblo committed:
Refactor throwing of exceptions

Methods would just call the print method instead of throwing
exceptions

Making methods throw exceptions and catching them at one place makes
it neater and more organized and predictable

Let's make methods throw exceptions instead of printing error messages

hernpiblo committed:
Create JAR file

hernpiblo committed:
Add find functionality

There is no find functionality

There will be better UX if user can search for tasks containing a
certain word

Let's add a find functionality to search for tasks

hernpiblo committed:
Add and modified Javadocs for almost all public methods

Public methods do not have proper documentation/javadocs after adding
and refactoring since last commit

Let's add proper javadocs for these public methods

hernpiblo committed:
Modify code to comply to coding standards

hernpiblo committed:
Merge branch 'branch-Level-9'

hernpiblo committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/pib/TaskList.java
#	src/main/java/pib/tasks/Deadline.java
#	src/main/java/pib/tasks/Event.java
#	src/main/java/pib/tasks/Task.java
#	src/main/java/pib/tasks/Todo.java

hernpiblo committed:
Merge branch 'branch-A-CodingStandard'

hernpiblo committed:
Modified storage to create new text file if it does not exist already

hernpiblo committed:
Merge remote-tracking branch 'origin/add-gradle-support'

hernpiblo committed:
Add Gradle to support build and test automation

hernpiblo committed:
Merge branch 'branch-A-Gradle'

hernpiblo committed:
Add GUI for Pib using JavaFX

hernpiblo committed:
Merge branch 'branch-Level-10'

benedictchuajj commented on src/main/java/pib/Pib.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public class Pib {
+
+    /**
+     * String to locate data file to load/save data
+     */
+    public static final String DATA_FILE_PATH = "data/data.txt";
```
I like how there's a comment here to help understand this variable 👍

benedictchuajj commented on src/main/java/pib/Pib.java:
```diff
@@ -0,0 +1,58 @@
+    public static final String DATA_FILE_PATH = "data/data.txt";
+
+    private Ui ui;
+    private TaskList list;
```
Perhaps consider changing the name to *tasks* so there's greater clarity of what this variable is supposed to be.

benedictchuajj commented on src/main/java/pib/gui/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+        setAlignment(Pos.TOP_LEFT);
+    }
+
+    public static DialogBox getUserDialog(String text, Image img) {
+        return new DialogBox(text, img);
+    }
+
+    public static DialogBox getPibDialog(String text, Image img) {
+        var db = new DialogBox(text, img);
+        db.flip();
+        return db;
+    }
```
Perhaps consider adding some Javadocs here since they are public methods? 😄

benedictchuajj commented on src/main/java/pib/gui/MainWindow.java:
```diff
@@ -0,0 +1,54 @@
+    private Image userImage = new Image(this.getClass().getResourceAsStream("/images/DaUser.png"));
+    private Image pibImage = new Image(this.getClass().getResourceAsStream("/images/DaDuke.png"));
+
+    @FXML
+    public void initialize() {
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    public void setPib(Pib p) {
+        pib = p;
+        dialogContainer.getChildren().add(DialogBox.getPibDialog(pib.loadData(), pibImage));
+    }
```
Small nit but perhaps you could add some Javadocs here too since they are public. Noticed the same issue in a few other places too.

benedictchuajj commented on src/main/java/pib/tasks/Deadline.java:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @return string format of Deadline task to be saved
+     */
+    public String toDataString() {
+        return "D," + getIsDone() + "," + getDescription() + "," + date + "," + time + System.lineSeparator();
+    }
```
Perhaps you can improve readability here by using String.format()? 
Noticed the same issues in several other places that returns a task as a String.

benedictchuajj commented on src/main/java/pib/utility/Parser.java:
```diff
@@ -0,0 +1,137 @@
+    /**
+     * Uses the scanner to read inputs from the user
+     */
+    public void readInput() {
+        scanner:
+        while (sc.hasNextLine()) {
+            try {
```
Any reason why you catch the exceptions in these Parser methods instead of throwing it to the main caller in Pib to catch? Since they all throw the same type of PibException, letting Pib catch these exceptions when it is calling these methods seems like a better control flow to me.

Just a nit from me and I might not be right about this 😃

benedictchuajj commented on src/main/java/pib/pibexception/PibException.java:
```diff
@@ -0,0 +1,29 @@
+     *
+     * @return String containing response to be printed to user
+     */
+    public String print() {
+        return Ui.printError(errorType);
+    }
```
I think this might be breaking the encapsulation of PibException as Ui knows the types of errorTypes and has to refer to it to return the errorMessage when it is done this way. Perhaps consider moving the switch statement from Ui to here to pass in only the errorMessage to Ui?

benedictchuajj reviewed (commented):
Looks good with very few coding standard violations. 👍 
Just a few nits to further improve readability.

Well done! I appreciate the effort you have put in to your Javadocs and comments throughout the code that makes it easy for me to follow. 😃

rldefa reviewed (commented):
Very clean code! Good job, no coding standard violation spotted. Very impressive : )

hernpiblo committed:
Add enableAssertions and modify mainClassName in build.gradle

hernpiblo committed:
Merge branch 'master' into branch-A-Assertions

hernpiblo committed:
Add assertions to some classes

No assertions are in the code.

Having assertions helps to minimise programming errors on dev's side
when we change something elsewhere in the future.

Let's add assertions to places where objects could be null and where
string could be null or blank.

hernpiblo committed:
Improve code quality

Code readability can be improved by modifying code structure

Better code quality helps to maintain the code in the future

Let's,
* implement SLAP
* refactor long methods and happy paths

hernpiblo committed:
Merge pull request #2 from hernpiblo/branch-A-CodeQuality

Improve code quality

hernpiblo committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/pib/Pib.java
#	src/main/java/pib/tasks/Event.java
#	src/main/java/pib/tasks/Todo.java
#	src/main/java/pib/utility/Storage.java
#	src/main/java/pib/utility/TaskList.java

hernpiblo committed:
Merge pull request #1 from hernpiblo/branch-A-Assertions

Branch a assertions

hernpiblo committed:
Add edit function to Pib

In order to edit a task, user have to delete and create a new task
with the updated values.

Having an edit function allows for better UX since user can just
edit the portion (description/date/time) of the task directly

Let's implement the edit function

hernpiblo committed:
Merge branch 'branch-BCD-Extension'

hernpiblo committed:
Add user guide and Ui screenshot

hernpiblo committed:
Update the correct README

hernpiblo committed:
Fix GFMD formatting issues

hernpiblo committed:
Improve look of User Guide

hernpiblo committed:
Update the javadocs based on bot comments

hernpiblo committed:
Fix issues from code review

hernpiblo committed:
Fix checkstyle issues

hernpiblo committed:
Improve code quality with more abstraction and fix checkstyle
issues
